tables/validate.py

```diff
@@ -0,0 +1,132 @@
+import re
+from dataclasses import dataclass
+from typing import Any
+
+import pandas as pd
+import pandera as pa
+
+from tables.table import Table
+
+
+@dataclass
+class Failure:
+    """Validation failure information returned by Pandera."""
+
+    schema_context: str
+    column: str
+    check: str
+    check_number: int
+    failure_case: str
+    index: int
+
+
+class ValidationError:
+    """Validation error with reference to an Excel cell."""
+
+    message: str
+    excel_cell_ref: str | None
+
+    def __init__(self, message: str, excel_cell_ref: str | None):
+        self.message = message
+        self.excel_cell_ref = excel_cell_ref
+
+
+class TableValidationError(Exception):
+    validation_errors: list[ValidationError]
+
+    def __init__(self, validation_errors: list[ValidationError]):
+        self.validation_errors = validation_errors
+
+
+class TableValidator:
+    """
+    Validates a table against a specified schema.
+
+    Attributes:
+        schema (pa.DataFrameSchema): The schema to validate against.
+        MESSAGE_OVERRIDE (dict[str, str]): A dictionary mapping validation failure messages
+            to custom override messages.
+        IGNORED_FAILURES (dict[str, str]): A dictionary mapping Failure attributes to regular expressions
+            for ignored validation failures. These regex patterns can be used to exclude
+            certain types of validation checks.
+
+    Example usage:
+    >>> schema_config = {
+    >>>     "columns": {
+    >>>         "name": pa.Column(str, checks=[pa.Check.str_length(max_value=50)]),
+    >>>         "age": pa.Column(int, checks=[pa.Check.greater_than(0)]),
+    >>>     }
+    >>> }
+    >>> validator = TableValidator(schema_config)
+    >>> table = Table(...)  # Assume you have a table instance
+    >>> try:
+    >>>     validator.validate(table)
+    >>> except TableValidationError as e:
+    >>>     for validation_error in e.validation_errors:
+    >>>         print(validation_error)
+    """
+
+    schema: pa.DataFrameSchema
+    MESSAGE_OVERRIDE = {
```

Comment:
Can see how this works and looks fairly clean, but just wonder if it's possible to control these errors manually through further implementation of custom checks to potentially avoid the need to use this? Creating a custom check to cover the `not_nullable` case seems on the face of it quite simple, although the other two slightly trickier given that they work across multiple values. So happy to keep this current logic if you aren't bothered

Reply:
You then also have to make sure that all `nullable` attributes are always set to `true` otherwise you'd still raise these errors. As this PR's scope has creeped a lot already I'm in favour of this to be done outside of this PR. Following work could look at creating predefined columns and checks that would pre-set a lot of our consistent behaviour (errors, checks, column params etc) that we can then use in any configs that we build rather than creating checks and columns from scratch in each config